src/java.base/share/classes/sun/nio/ch/SocketChannelImpl.java

```diff
@@ -134,48 +136,35 @@
         super(sp);
         Objects.requireNonNull(family, "'family' is null");
         if ((family != StandardProtocolFamily.INET) &&
+                (family != StandardProtocolFamily.UNIX) &&
                 (family != StandardProtocolFamily.INET6)) {
             throw new UnsupportedOperationException("Protocol family not supported");
         }
-        if (family == StandardProtocolFamily.INET6 && !Net.isIPv6Available()) {
+        if (family == StandardProtocolFamily.UNIX) {
+            this.fd = UnixDomainSockets.socket();
+        } else if (family == StandardProtocolFamily.INET6 && !Net.isIPv6Available()) {
             throw new UnsupportedOperationException("IPv6 not available");
+        } else {
+            this.fd = Net.socket(family, true);
         }
         this.family = family;
-        this.fd = Net.socket(family, true);
         this.fdVal = IOUtil.fdVal(fd);
     }
```

Comment:
If you introduced:
```
private static FileDescriptor socketFor(ProtocolFamily family) throws IOException ...
```
and 
```
private SocketChannelImpl(SelectorProvider sp, ProtocolFamily family, FileDescriptor fd) throws IOException {
      super(sp);
      this.family = family;
      this.fd = fd;
      this.fdVal = IOUtil.fdVal(fd);
 }
```
then you could move all the checking to `socketFor(ProtocolFamilly)` and write this constructor as:
```
SocketChannelImpl(SelectorProvider sp, ProtocolFamily family) throws IOException {
    this(sp, family, sockectFor(family));
}
```
which would push all the validation to **before** the object gets constructed.